delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala

```diff
@@ -0,0 +1,1166 @@
+        notMatchedOutputs = notMatchedOutputs,
+        notMatchedBySourceConditions = cmd.notMatchedBySourceClauses.map(clauseCondition),
+        notMatchedBySourceOutputs = cmd.notMatchedBySourceClauses.map(clauseOutput),
+        joinedAttributes = joinedPlan.output,
+        joinedRowEncoder = joinedRowEncoder,
+        processedRowEncoder = processedRowEncoder)
+
+      Dataset.ofRows(spark, joinedPlan)
+        .mapPartitions(processor.processPartition)(processedRowEncoder)
+    }
+
+    private def addRepartitionByFilePath(input: DataFrame): DataFrame = {
+      logDebug(s"Trying to adding RapidsRepartitionByFilePath plan node")
+      val newLogicalPlan = input.queryExecution.optimizedPlan.transformUp({
+        case p@Project(projectList, r @ LogicalRelation(fs: HadoopFsRelation, _, _, _)) =>
+          if (projectList.exists(_.name == FILE_PATH_COL)) {
+            // Here we enforce that the targetDF to read each file per partition. This is to ensure
+            // that we can avoid shuffling the target data when doing left anti join in the second
+            // step of low shuffle merge. See
+            // [[com.nvidia.spark.rapids.delta .RapidsRepartitionByFilePathExec]] for more details.
+            val newOptions = fs.options + (RapidsConf.FORCE_ONE_FILE_PER_PARITION.key -> "true")
+            val newFs = fs.copy(options = newOptions)(spark)
+            val newRelation = r.copy(relation = newFs)
+
+            val newProjection = p.copy(child = newRelation)
+            RapidsRepartitionByFilePath(None, newProjection)
+          } else {
+            p
+          }
+      })
+
+      Dataset.ofRows(spark, newLogicalPlan)
+    }
+
+    private def makeUnmodifiedTargetDF(
+        baseTargetDF: DataFrame,
+        modifiedTargetRows: DataFrame): DataFrame = {
+      // We make the join condition as following:
+      // left._metadata_file_path = right._metadata_file_path &&
+      //  (left._metadata_row_id + right._metadata_row_id) = 0
+      // to ensure that the generated physical plan only requires repartition
+      // by `_metadata_file_path`, so that we can avoid shuffle of target table.
+      val joinCondition = baseTargetDF(FILE_PATH_COL) === modifiedTargetRows(FILE_PATH_COL) &&
+        ((baseTargetDF(METADATA_ROW_IDX_COL) +
+          modifiedTargetRows(METADATA_ROW_IDX_COL)) === 0L)
+
+
+      baseTargetDF.join(modifiedTargetRows,
+          joinCondition, "anti")
+        .drop(METADATA_ROW_IDX_COL, FILE_PATH_COL)
+    }
+
+    private def checkGpuFileSourceScanExecOverride(input: DataFrame): Boolean = {
+        val executedPlan = input.queryExecution
+          .executedPlan
+
+        var pushed = false
+        GpuOverrides().applyWithContext(executedPlan.clone(), Some(
+          """Detecting GpuFileSourceScanExec could be pushed down""".stripMargin)).foreachUp {
+          case GpuRapidsRepartitionByFilePathExec(_, _,
+          GpuProjectExec(_, _: GpuFileSourceScanExec)) =>
```

Comment:
This seems very specific and fragile.  Aren't there usually filters inserted by Spark immediately after a scan to remove nulls before a join, for example?  We don't actually care if there's a row-to-columnar or columnar-to-row in here, we just need there to be a GpuFileSourceScanExec underneath the GpuRapidsRepartitionByFilePathExec without partitioning changes in-between, correct?